### PR TITLE
Handling of long running PVC creation from snapshot request

### PR DESCRIPTION
**Pull request by smitaraut**
During a volume creation request from snapshot, if the snapshot has millions of files, the copy operation from snapshot to the new volume takes very long causing the context deadline to be exceeded. In such case kubernetes retries the request and CSI driver was again initiating the snapshot copy, and the cycle of context deadline exceeded would continue causing the PVC to remain in pending state forever.

**Fix:**
A map is maintained in the CSI driver for PVC Id to the status of snapshot copy job. In case of request retry, the CSI driver checks if a snapshot copy request for the same PVC is already in progress or has succeeded/failed and returns appropriate response. 

**Tests:**
Tested with 1 million files snapshot restore, multiple simultaneous restores.

### CODE REVIEW

File: driver/csiplugin/connectors/rest_v2.go [hunk 1]
```diff
@@ -278,18 +278,24 @@ func (s *spectrumRestV2) CopyFsetSnapshotPath(filesystemName string, filesetName
...
+	return copySnapResp.Status.Code, copySnapResp.Jobs[0].JobID, nil
+}
+
+func (s *spectrumRestV2) WaitForSnapshotCopy(statusCode int, jobID uint64) error {
+	glog.V(4).Infof("rest_v2 WaitForSnapshotCopy. statusCode: %v, jobID: %v", statusCode, jobID)
+
+	err := s.waitForJobCompletion(statusCode, jobID)
 	if err != nil {
-		glog.Errorf("Unable to copy snapshot %s: %v", snapshotName, err)
+		glog.Errorf("Error in waiting for job completion %v, %v", jobID, err)
```
Comment (deeghuge) [inline driver/csiplugin/connectors/rest_v2.go:298]: error strings should not be capitalized

File: driver/csiplugin/controllerserver.go [hunk 2]
```diff
@@ -518,6 +517,25 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
 		return nil, status.Error(codes.Aborted, fmt.Sprintf("volume creation already in process : %v", scaleVol.VolName))
 	}
 
+	jobDetails := cs.Driver.snapjobstatusmap[scaleVol.VolName]
+	if jobDetails.jobStatus == SNAP_JOB_RUNNING {
+		glog.Errorf("volume:[%v] -  Snapshot copy request in progress for snapshot: %s.", scaleVol.VolName, snapIdMembers.SnapName)
+		return nil, status.Error(codes.Aborted, fmt.Sprintf("Snapshot copy request in progress for snapshot: %s", snapIdMembers.SnapName))
+	} else if jobDetails.jobStatus == SNAP_JOB_FAILED {
+		glog.Errorf("volume:[%v] -  Snapshot copy job had failed for snapshot %s", scaleVol.VolName, snapIdMembers.SnapName)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("SSnapshot copy job had failed for snapshot: %s", snapIdMembers.SnapName))
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:526]: error strings should not be capitalized and there is typo SSnapshot

File: driver/csiplugin/controllerserver.go [hunk 3]
```diff
@@ -550,9 +568,9 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
 	}
 
 	if isSnapSource {
-		err = cs.copySnapContent(scaleVol, snapIdMembers, fsDetails, targetPath)
+		err = cs.copySnapContent(scaleVol, snapIdMembers, fsDetails, targetPath, volID)
 		if err != nil {
-			glog.Errorf("CreateVolume [%s]: [%v]", volName, err)
+			glog.Errorf("CreateVolume failed while copying snapshot content [%s]: [%v]", volName, err)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:573]: error strings should not be capitalized

File: driver/csiplugin/controllerserver.go [hunk 4]
```diff
@@ -582,13 +600,29 @@ func (cs *ScaleControllerServer) copySnapContent(scVol *scaleVolume, snapId scal
...
+	err = conn.WaitForSnapshotCopy(jobStatus, jobID)
+	if err != nil {
+		glog.Errorf("Unable to copy snapshot %s: %v.", snapId.SnapName, err)
+		jobDetails.jobStatus = SNAP_JOB_FAILED
+		cs.Driver.snapjobstatusmap[scVol.VolName] = jobDetails
+		//delete(cs.Driver.snapjobmap, scVol.VolName)
+		return err
+	}
+
+	glog.Infof("Copy snapshot completed for snapId: [%v], scaleVolume: [%v]", snapId, scVol)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:622]: error strings should not be capitalized

File: driver/csiplugin/controllerserver.go [hunk 5]
```diff
@@ -805,6 +839,11 @@ func (cs *ScaleControllerServer) DeleteVolume(ctx context.Context, req *csi.Dele
 				//Check if fileset exist has any snapshot
 				snapshotList, err := conn.ListFilesetSnapshots(FilesystemName, FilesetName)
 				if err != nil {
+					if strings.Contains(err.Error(), "EFSSG0072C") ||
+						strings.Contains(err.Error(), "400 Invalid value in 'filesetName'") { // fileset is already deleted
+						glog.V(4).Infof("Fileset seems already deleted - %v", err)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:844]: error strings should not be capitalized

File: driver/csiplugin/controllerserver.go [hunk 6]
```diff
@@ -815,6 +854,11 @@ func (cs *ScaleControllerServer) DeleteVolume(ctx context.Context, req *csi.Dele
 
 				err = conn.DeleteFileset(FilesystemName, FilesetName)
 				if err != nil {
+					if strings.Contains(err.Error(), "EFSSG0072C") ||
+						strings.Contains(err.Error(), "400 Invalid value in 'filesetName'") { // fileset is already deleted
+						glog.V(4).Infof("Fileset seems already deleted - %v", err)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:859]: error strings should not be capitalized

**Review by deeghuge — COMMENTED** [review]
- mostly the log message related comments.
- Since we are not removing the entry from snapshot map so I don't think it will grow that much but worth a check
- and as you already mentioned driver pod restart case(where snapshot map gets deleted) will be taken care in another issue.

File: driver/csiplugin/connectors/rest_v2.go (hunk 1 above)
Comment (smitaraut) [reply driver/csiplugin/connectors/rest_v2.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 2 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 3 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 4 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 5 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 6 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: Done

**Review by deeghuge — APPROVED** [review]
lgtm

**Review by Jainbrt — APPROVED** [review]
LGTM

**Pull request merged**